lib/Optimizer/Transforms/GenKernelExecution.cpp

```diff
@@ -1106,6 +1108,37 @@ class GenerateKernelExecution
         continue;
       }
       if (auto strTy = dyn_cast<cudaq::cc::StructType>(quakeTy)) {
+        if (!isa<cudaq::cc::PointerType>(arg.getType())) {
+          // If argument is not a pointer, then struct was promoted into a
+          // register.
+          auto *parent = builder.getBlock()->getParentOp();
+          auto module = parent->getParentOfType<ModuleOp>();
+          auto tmp = builder.create<cudaq::cc::AllocaOp>(loc, quakeTy);
+          auto cast = builder.create<cudaq::cc::CastOp>(
+              loc, cudaq::cc::PointerType::get(arg.getType()), tmp);
+          if (cudaq::opt::factory::isX86_64(module)) {
+            builder.create<cudaq::cc::StoreOp>(loc, arg, cast);
+            if (cudaq::opt::factory::structUsesTwoArguments(quakeTy)) {
+              auto cast = builder.create<cudaq::cc::CastOp>(
+                  loc, cudaq::cc::PointerType::get(builder.getI8Type()), tmp);
+              auto hiPtr = builder.create<cudaq::cc::ComputePtrOp>(
+                  loc, cudaq::cc::PointerType::get(builder.getI8Type()), cast,
+                  cudaq::cc::ComputePtrArg{8});
+              ++iter;
+              Value nextArg = *iter;
```

Comment:
I'm curious why `iter` would be incremented but `idx` would not?

Reply:
The purpose of that change and the others in this loop is that there are more physical arguments, potentially, than logical arguments. This is the case where a logical argument (idx) is 2 physical arguments (iter).